### PR TITLE
TASK_zfyujmqt Create generic sqlexecutabledriver interface for sqlexecutable class with tests + refactor planetscale sqlexecutable

### PR DESCRIPTION
**Pull request by suryatiru**
Problem: Currently our SQLExecutable is coupled with Planetscale’s js package and has no isolated tests. If we were to switch to a different provider or multiple providers, it would be difficult to determine if the migration works without running the whole test suite.

Solution: Create a SQLExecutableDriver file that implements .query, .queryResult and .transaction. We can make separate versions of SQLExecutableDriver for our local mySQL provider. Create isolated tests for SQLExecutableDriver that validate the behavior of query, queryResult and transaction and validate the data type transformations from MySQL to typescript using a test table.

Deliverables:

- SQLExecutableDriver implementation for local mySQL's js package.
- tests for execute, `executeAndReturnQueryResult` and `transaction` methods.
  - In tests, validate query variables work and validate query result data types (string, boolean, number).

Ticket ID: TASK_zfyujmqt

### CODE REVIEW

**Comment by tifapp** [comment]
Linked Ticket: https://trello.com/c/zfyujmqt

File: TiFBackendUtils/SQLExecutable/utils.ts [hunk 1]
```diff
@@ -174,8 +174,3 @@ export class SQLExecutable {
     )
   }
 }
-
-// Usage
```
Comment (seanim0920) [inline TiFBackendUtils/SQLExecutable/utils.ts:178]: We could keep the comments

File: TiFBackendUtils/SQLExecutable/index.ts [hunk 2]
```diff
@@ -57,4 +57,4 @@ export const recreateDatabase = async () => {
   } catch (error) {
     console.error('An error occurred:', error);
   }
-};
```
Comment (seanim0920) [inline TiFBackendUtils/SQLExecutable/index.ts:60]: Duplicate code can be removed

File: TiFBackendUtils/LocalSQL/MySQLDriver.ts [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+  async query<Value> (
+    query: string,
+    args: object | null
+  ): Promise<Value[]> {
+    // Use this.conn to execute the query and return the result rows
+    // This will be the only function to directly use the database library's execute method.
+    const conn = await this.conn
+    const [rows] = await conn.query(query, args)
+    if (Array.isArray(rows)) {
+      return castTypes(rows as RowDataPacket[]) as Value[]
```
Comment (ZoroarkZark) [inline TiFBackendUtils/LocalSQL/MySQLDriver.ts]: Why do we need to cast types here, and specifically like this? I'm not sure I follow the reasoning here

File: TiFBackendUtils/LocalSQL/MySQLDriver.ts [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+   *
+   * @example
+   * ```ts
+   * type User = { id: string, ... }
+   *
+   * const result? = await queryFirst<User>(conn, "SELECT * FROM user");
+   * console.log(result?.id) // ✅ Typesafe
+   * ```
+   */
+  queryFirstResult (query: string, args: object | null = null) {
```
Comment (ZoroarkZark) [inline TiFBackendUtils/LocalSQL/MySQLDriver.ts]: Why would we need a method for just the first result specifically? It seems like it's not very flexible

File: TiFBackendUtils/SQLExecutable/tests/MySQLDriver.test.ts [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    PRIMARY KEY (id)
+    )`
+    await conn.transaction(
+      async (tx) => await tx.executeResult(createMySQLDriverTableSQL)
+    )
+  })
+
+  afterAll(async () => {
+    const deleteMySQLDriverTableSQL = `
+    DROP TABLE IF EXISTS MySQLDriver
```
Comment (ZoroarkZark) [inline TiFBackendUtils/SQLExecutable/tests/MySQLDriver.test.ts]:
```suggestion
    DROP TABLE IF EXISTS mySQLDriver
```

File: TiFBackendUtils/LocalSQL/MySQLDriver.ts (hunk 3 above)
Comment (suryatiru) [reply TiFBackendUtils/LocalSQL/MySQLDriver.ts]: When converting the mySQL values to say a boolean or number we use the castTypes function. Maybe there could be a better way to do that functionality.

File: TiFBackendUtils/LocalSQL/dbConnection.ts [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    })
+    console.log("Successfully connected to the database.")
+    return connection
+  } catch (error) {
+    console.error("Unable to connect to the database:", error)
+    throw error
+  }
+}
+
+export const sqlConn = createDatabaseConnection()
```
Comment (seanim0920) [inline TiFBackendUtils/LocalSQL/dbConnection.ts]: Do we use sqlConn anywhere or can we merge this with the next line

**Comment by seanim0920** [comment]
Can you rename the folder LocalSQL to MySQLDriver

File: TiFBackendUtils/LocalSQL/dbConnection.ts (hunk 6 above)
Comment (suryatiru) [reply TiFBackendUtils/LocalSQL/dbConnection.ts]: Yes, we use that connection in the LocalSQL file to reset the database.

**Review by seanim0920 — APPROVED** [review]

**Pull request merged**